plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+member' k = go
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        False
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then True
+                  else go tl
+        )
+
+{-# INLINEABLE insert #-}
+insert :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a -> Map k a
+insert (P.toBuiltinData -> k) (P.toBuiltinData -> a) m =
+  unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        (BI.mkCons (BI.mkPairData k a) nil)
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then BI.mkCons (BI.mkPairData k a) tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE delete #-}
+delete :: forall k a. (P.ToData k) => k -> Map k a -> Map k a
+delete (P.toBuiltinData -> k) m = unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        nil
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE singleton #-}
+singleton :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a
+singleton (P.toBuiltinData -> k) (P.toBuiltinData -> a) = unsafeFromBuiltinList xs
+  where
+    xs = BI.mkCons (BI.mkPairData k a) nil
+
+{-# INLINEABLE empty #-}
+empty :: forall k a. Map k a
+empty = unsafeFromBuiltinList nil
+
+{-# INLINEABLE null #-}
+null :: forall k a. Map k a -> Bool
+null = P.null . toBuiltinList
+
+{-# INLINEABLE unsafeFromList #-}
+unsafeFromList :: (P.ToData k, P.ToData a) => [(k, a)] -> Map k a
+unsafeFromList =
+  unsafeFromBuiltinList
+    . toBuiltin
+    . PlutusTx.Prelude.map (\(k, a) -> (P.toBuiltinData k, P.toBuiltinData a))
+
+{-# INLINEABLE uncons #-}
+uncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  Map k a ->
+  Maybe ((k, a), Map k a)
+uncons m = case P.uncons (toBuiltinList m) of
+  Nothing -> Nothing
+  Just (pair, rest) ->
+    let (k, a) = P.pairToPair pair
+     in Just ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList rest)
+
+{-# INLINEABLE unsafeUncons #-}
+unsafeUncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  Map k a ->
+  ((k, a), Map k a)
+unsafeUncons m =
+  ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList tl)
+  where
+    (hd, tl) = P.unsafeUncons (toBuiltinList m)
+    (k, a) = P.pairToPair hd
+
+{-# INLINEABLE noDuplicateKeys #-}
+noDuplicateKeys :: forall k a. Map k a -> Bool
```

Comment:
this is quadratic, worth saying

Reply:
Again, if we call this an association list, then this would be a lot more obvious.